spark-operator-api/src/main/java/org/apache/spark/kubernetes/operator/status/ApplicationStateSummary.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.spark.kubernetes.operator.status;
+
+import java.util.Set;
+
+public enum ApplicationStateSummary implements BaseStateSummary {
+  /** Spark application is submitted to the cluster but yet scheduled. */
+  SUBMITTED,
+
+  /** Spark application will be restarted with same configuration */
+  SCHEDULED_TO_RESTART,
+
+  /** A request has been made to start driver pod in the cluster */
+  DRIVER_REQUESTED,
+
+  /** Driver pod has reached running state */
+  DRIVER_STARTED,
+
+  /** Spark session is initialized */
+  DRIVER_READY,
+
+  /** Less that minimal required executor pods become ready during starting up */
```

Comment:
I don't think this description is correct. At least, this is insufficient.
- What is the meaning of `Executor Pod Become Ready`?
- Executor Pod readiness doesn't imply that Driver JVM knows this executor.

Reply:
Updated description. You are absolutely right that executor state does not imply their actual state from Spark perspective. This is a 'best effort' from operator side to observe app status without modification to the core.

We do have some ideas to optimize this in future versions. Making operator able to detect app status by:

* connect to driver to get its registered executor information (instead of watching executor pods). I may use existing SparkUI for this purpose - user should be able to opt-in this feature if they enables pod-to-pod communication between operator and driver.
* have the driver updates CRD status, possibly via a listener. 

These future enhancement may involve core / k8s module changes.